### PR TITLE
UCD-94: Add audio support for Core Desktop

### PR DESCRIPTION
**Pull request by sergio-costas**
This is one of the three MRs required for adding audio support to Core Desktop. This one adds the last version of PipeWire, which will be needed to add support of permissions. It also adds some extra packages needed for gnome-control-center to work as expected (for example, without libcanberra-pulse, the audio test doesn't work).

### CODE REVIEW

**Review by jhenstridge — CHANGES_REQUESTED** [review]
Sorry for the delay in getting back to you on this. There are a few things to consider:

1. if we're shipping a modified version of pipewire, would it be easier to manage this as a deb package? The base snap build is already set up to pull packages from the [desktop-snappers PPA](https://launchpad.net/~desktop-snappers/+archive/ubuntu/core-desktop). This would remove the possibility of accidentally mixing binaries or data files from the two releases, and make dependency conflicts more obvious.
2. I think we want to have pipewire + friends running in the context of the ubuntu-desktop-session snap. This'd be expected if that snap is providing the relevant slots, and will be needed for e.g. xdg-desktop-portal-gnome + gnome-shell to talk to pipewire to set up screen sharing. This would probably mean stripping out the systemd units that start pipewire, and add equivalent user daemons to ubuntu-desktop-session to start them. We'll also likely need to add at least an `alsa` plug to ubuntu-desktop-session to give it access to the audio devices.
3. We'll need some changes to the pipewire snap support for this mode, since we don't have full access to the snapd API in that context. We do have a snapctl interface that does what we need (but annoyingly can't replace the code we use for an unconfined pipewire/pulse):
    ```
    snapctl is-connected --pid=$PID audio-record
    snapctl is-connected --apparmor-label=$LABEL audio-record
    ```
   ... passing the pid or label of the connection peer. If the exit code is not 1, then we should act as if they have audio-record connected. We can either shell out to `snapctl`, or use `libsnapd-glib`'s [snapctl API](https://github.com/snapcore/snapd-glib/blob/master/examples/snapctl-glib.c) here.

**Comment by sergio-costas** [comment]
I see a problem with pipewire running inside the context of ubuntu-desktop-session: in that case, the pipewire sockets would be created at _/run/user/XXXX/snap.ubuntu-desktop-session_ instead of _/run/user/XXXX_, and so, only the programs running inside that context would have access to it. Giving to it audio permissions would allow to create the _pulse_ sockets at _/run/user/XXXX_, but what happens with _pipewire-0_ and _pipewire-0.lock_?

**Comment by jhenstridge** [comment]
I might be mistaken, but my understanding is that confined apps generally access pipewire via file descriptors passed to them via xdg-desktop-portal. In this case it'll be gnome-shell and xdg-desktop-portal-gnome (which is running inside the ubuntu-desktop-session snap) that establish these connections, so they should see the pipewire socket in the private XDG_RUNTIME_DIR just fine.

**Comment by sergio-costas** [comment]
I can test it...

**Comment by sergio-costas** [comment]
Ok, I just removed the systemd files that launched pipewire at user login, and launch it manually in the session script, so it runs inside the session. It works fine.

**Comment by sergio-costas** [comment]
The odd thing is that I didn't need to connect the ALSA interface... but pipewire is running inside the session because the pipewire-0 and pipewire-0.lock files are inside /run/user/1001/snap.ubuntu-desktop-session...

**Comment by jhenstridge** [comment]
> Ok, I just removed the systemd files that launched pipewire at user login, and launch it manually in the session script, so it runs inside the session. It works fine.

My idea was that we would treat this as a user daemon. I think this'd be the appropriate snapcraft.yaml syntax:

```yaml
apps:
    ...
    pipewire:
      command: run.sh /usr/bin/pipewire
      daemon: simple
      passthrough:
        daemon-scope: user
      sockets:
        pipewire:
          listen-stream: $XDG_RUNTIME_DIR/pipewire-0
      plugs:
       ...
       - network-bind
```

This would give us a socket activated user daemon, similar to how it is launched normally. We'd need similar ones for pipewire-pulse and wireplumber.

What we would need in the setup script is something to create a symlink from the real `$XDG_RUNTIME_DIR/pulse` to the private dir to expose Pulse Audio protocol to other snaps.

**Comment by sergio-costas** [comment]
About the dektop-snappers PPA, how can I send a package there?

**Comment by sergio-costas** [comment]
The symlink is not needed as long as https://github.com/canonical/ubuntu-desktop-session-snap/pull/1/files is run before, because it sets PULSE_SERVER environment variable to the global path (accessible because the snap has audio-playback/audio-record privileges), ensuring that all other snaps can access it.

**Comment by sergio-costas** [comment]
Mmm... if we use "run.sh" to launch them, we can add the EXPORT command there...

**Comment by jhenstridge** [comment]
The main benefits of treating them as user daemons are:
1. it's closer to how classic Ubuntu treats them
2. we can set a restart policy to make the services more durable
3. socket activation solves some of the startup ordering problems, since the sockets will be created early on and anything that wants to talk to the services will block until they start up
4. their log messages will automatically be collected, and correctly categorised

**Comment by sergio-costas** [comment]
Yes, fully agree. I meant that in the example piece that you put to launch pipewire as a daemon, you don't call it directly, but using "run.sh" instead, and that help us to solve the problem of the path for the socket.

**Comment by jhenstridge** [comment]
The use of `run.sh` script is there to work around the fact that Snapcraft wants the command for an app or daemon to be a file within the snap itself. It's just there to let us use binaries provided by the base snap as commands/daemons.

File: hooks/000-provide-uids-gids.chroot
Comment (jhenstridge) [inline hooks/000-provide-uids-gids.chroot:125]: Could you set this uid to 115? Previously 107 was allocated to the docker user (see 45d8020), and it seems safer not to reuse uids from core22.

It's not clear whether we'll encounter file systems using Ubuntu Core's uids/gids, but it seems better not to risk it.

File: hooks/000-provide-uids-gids.chroot
Comment (jhenstridge) [inline hooks/000-provide-uids-gids.chroot:231]: Same for reuse of gid 113 here. That was previously allocated to docker.

File: hooks/990-ensure-uids-gids.chroot
Comment (jhenstridge) [inline hooks/990-ensure-uids-gids.chroot:29]: We can leave these tests as is if we don't reuse these uids/gids.

File: snapcraft.yaml
Comment (jhenstridge) [inline snapcraft.yaml:100]: Can you move this to `hooks/006-add-gdm.chroot`, where we perform similar fixes?

File: hooks/001-zz-add-core-desktop-ppa.chroot
Comment (jhenstridge) [inline hooks/001-zz-add-core-desktop-ppa.chroot:58]: When you're happy with these packages, it'd be good to move them to the desktop-snappers PPA.

File: snapcraft.yaml
Comment (jhenstridge) [inline snapcraft.yaml:84]: Please don't add unnecessary white-space changes. It increases the chance of merge conflicts when we pull in changes from the core-base repo.

**Review by jhenstridge — CHANGES_REQUESTED** [review]
Overall, this looks pretty close to ready. I've left a few comments below.

File: snapcraft.yaml
Comment (sergio-costas) [reply snapcraft.yaml]: fixed.

File: hooks/001-zz-add-core-desktop-ppa.chroot
Comment (sergio-costas) [reply hooks/001-zz-add-core-desktop-ppa.chroot:58]: I prefer to wait until this is merged, to ensure that there are no conflicts.

File: hooks/000-provide-uids-gids.chroot
Comment (sergio-costas) [reply hooks/000-provide-uids-gids.chroot:125]: Done.

File: hooks/990-ensure-uids-gids.chroot
Comment (sergio-costas) [reply hooks/990-ensure-uids-gids.chroot:29]: Done.

Comment (sergio-costas) [reply hooks/990-ensure-uids-gids.chroot:29]: Done.

File: hooks/000-provide-uids-gids.chroot
Comment (sergio-costas) [reply hooks/000-provide-uids-gids.chroot]: Done.

File: hooks/001-zz-add-core-desktop-ppa.chroot
Comment (kenvandine) [inline hooks/001-zz-add-core-desktop-ppa.chroot:58]: I'd rather not enable yet another PPA, it would be better to copy the pipewire package to the ~desktop-snappers/core-desktop PPA

Comment (sergio-costas) [reply hooks/001-zz-add-core-desktop-ppa.chroot]: Fixed.

**Pull request merged**
